main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+    case ResolvedAst.Predicate.Body.Atom(pred, den, polarity, terms0, loc) =>
+      for {
+        terms <- Validation.traverse(terms0)(ascribePattern(_, kenv, root))
+      } yield KindedAst.Predicate.Body.Atom(pred, den, polarity, terms, Type.freshVar(Kind.Star), loc)
+    case ResolvedAst.Predicate.Body.Guard(exp0, loc) =>
+      for {
+        exp <- ascribeExpression(exp0, kenv, root)
+      } yield KindedAst.Predicate.Body.Guard(exp, loc)
+  }
+
+  /**
+    * Infers a kind environment from the given spec.
+    */
+  private def inferSpec(spec0: ResolvedAst.Spec, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = spec0 match {
+    case ResolvedAst.Spec(_, _, _, _, fparams, sc, _, eff, _) =>
+      val fparamKenvVal = Validation.fold(fparams, KindEnv.empty) {
+        case (acc, fparam) => flatMapN(inferFparam(fparam, root)) {
+          fparamKenv => acc ++ fparamKenv
+        }
+      }
+      val schemeKenvVal = inferScheme(sc, root)
+      val effKenvVal = inferType(eff, KindMatch.subKindOf(Kind.Bool), root)
+
+      flatMapN(fparamKenvVal, schemeKenvVal, effKenvVal) {
+        case (fparamKenv, schemeKenv, effKenv) => KindEnv.merge(fparamKenv, schemeKenv, effKenv)
+      }
+
+  }
+
+  /**
+    * Infers a kind environment from the given formal param.
+    */
+  private def inferFparam(fparam0: ResolvedAst.FormalParam, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = fparam0 match {
+    case ResolvedAst.FormalParam(_, _, tpe0, _) => inferType(tpe0, KindMatch.subKindOf(Kind.Star), root)
+  }
+
+  /**
+    * Infers a kind environment from the given scheme.
+    */
+  private def inferScheme(sc0: ResolvedAst.Scheme, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = sc0 match {
+    case ResolvedAst.Scheme(_, constraints, base) =>
+      val baseKenvVal = inferType(base, KindMatch.subKindOf(Kind.Star), root)
+      val tconstrsKenvsVal = Validation.traverse(constraints)(inferTconstr(_, root))
+
+      Validation.flatMapN(baseKenvVal, tconstrsKenvsVal) {
+        case (baseKenv, tconstrKenvs) => Validation.fold(tconstrKenvs, baseKenv)(_ ++ _)
+      }
+  }
+
+  /**
+    * Infers a kind environment from the given type constraint.
+    */
+  private def inferTconstr(tconstr: ResolvedAst.TypeConstraint, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = tconstr match {
+    case ResolvedAst.TypeConstraint(clazz, tpe, loc) =>
+      val kind = getClassKind(root.classes(clazz))
+      inferType(tpe, KindMatch.subKindOf(kind), root)
+  }
+
+  /**
+    * Infers a kind environment from the given type, with an expectation from context.
+    */
+  private def inferType(tpe: UnkindedType, expectedType: KindMatch, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindEnv, KindError] = tpe.baseType match {
```

Comment:
Whats the difference between the ascribeX and inferX? I think its that one translates and the other gives back a kind env.

Reply:
Yes, that's about right. `inferX` builds a kind environment and `ascribeX` uses the kind environment to ascribe kinds to the types.